File: Dataset/Parser.py
import csv
from datetime import datetime

class Position:
    def __init__(self, time, latitude, longitude, altitude, groundSpeed, idAircraft, modelAircraft):
        self.time = time
        self.latitude = latitude
        self.longitude = longitude
        self.altitude = altitude
        self.groundSpeed = groundSpeed
        self.idAircraft = idAircraft
        self.modelAircraft = modelAircraft

    def __str__(self):
        return f"Position{{time='{self.time}', latitude={self.latitude}, longitude={self.longitude}, altitude={self.altitude}, groundSpeed={self.groundSpeed}, idAircraft='{self.idAircraft}', modelAircraft='{self.modelAircraft}'}}"

class Flight:
    def __init__(self, saida, chegada):
        self.saida = saida
        self.chegada = chegada

    def __str__(self):
        return f"Flight [saida={self.saida}, chegada={self.chegada}]"

def is_interval_greater_than_three_minutes(time1, time2):
    try:
        date1 = datetime.strptime(time1, '%Y-%m-%d %H:%M:%S.%f')
        date2 = datetime.strptime(time2, '%Y-%m-%d %H:%M:%S.%f')
        diff_in_minutes = abs((date2 - date1).total_seconds() / 60)
        return diff_in_minutes > 3
    except Exception as e:
        print(f"Erro ao comparar datas: {e}")
        return False

def parse(filename):
    positions = []
    voos = []

    try:
        with open(filename, newline='', encoding='utf-8') as csvfile:
            reader = csv.reader(csvfile)
            next(reader)  # Ler e descartar o cabeçalho

            for row in reader:
                current_position = Position(
                    time=row[0],
                    latitude=float(row[1]),
                    longitude=float(row[2]),
                    altitude=float(row[3]),
                    groundSpeed=float(row[4]),
                    idAircraft=row[5],
                    modelAircraft=row[6]
                )
                positions.append(current_position)

        # Processar as posições para segmentar voos
        i = 0
        while i < len(positions):
            start = positions[i]
            end = start
            j = i + 1
            while j < len(positions):
                next_pos = positions[j]
                if is_interval_greater_than_three_minutes(end.time, next_pos.time):
                    break
                end = next_pos
                j += 1
            if end != start:
                flight = Flight(saida=start.time, chegada=end.time)
                voos.append(flight)
                i = j  # Avançar o ponteiro para o próximo segmento de voo
            else:
                i += 1

    except Exception as e:
        print(e)

    print("Voos segmentados:")
    for voo in voos:
        print(voo)

if __name__ == "__main__":
    sorted_positions_csv = "E:/Projetos/TesteTecnico/Dataset/PosSorted/11.csv"
    parse(sorted_positions_csv)

File: Dataset/Parser.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;

public class Parser2 {
    public static void main(String[] args) {
        String sortedPositionsCsv = "E:/Projetos/TesteTecnico/Dataset/PosSorted/11.csv";
        parse(sortedPositionsCsv);
    }

    public static void parse(String filename) {
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");
        ArrayList<Position> positions = new ArrayList<>();
        ArrayList<Flight> voos = new ArrayList<>();
        
        try {
            BufferedReader reader = new BufferedReader(new FileReader(filename));
            String line = reader.readLine(); // Ler e descartar o cabeçalho
            
            while ((line = reader.readLine()) != null) {
                String[] fields = line.split(",");
                Position currentPosition = new Position();
                currentPosition.time = fields[0];
                currentPosition.latitude = Double.parseDouble(fields[1]);
                currentPosition.longitude = Double.parseDouble(fields[2]);
                currentPosition.altitude = Double.parseDouble(fields[3]);
                currentPosition.groundSpeed = Double.parseDouble(fields[4]);
                currentPosition.idAircraft = fields[5];
                currentPosition.modelAircraft = fields[6];
                positions.add(currentPosition);
            }
            
            reader.close();
            
            // Processar as posições para segmentar voos
            for (int i = 0; i < positions.size(); i++) {
                Position start = positions.get(i);
                Position end = start;
                for (int j = i + 1; j < positions.size(); j++) {
                    Position next = positions.get(j);
                    if (isIntervalGreaterThanThreeMinutes(sdf, end.time, next.time)) {
                        break;
                    } else {
                        end = next;
                    }
                }
                if (end != start) {
                    Flight flight = new Flight();
                    flight.saida = start.time;
                    flight.chegada = end.time;
                    voos.add(flight);
                    // Avançar o ponteiro de análise
                    i = positions.indexOf(end);
                }
            }
            
        } catch (IOException e) {
            e.printStackTrace();
        }
        
        System.out.println("Voos segmentados:");
        for (Flight voo : voos) {
            System.out.println(voo);
        }
    }

    public static boolean isIntervalGreaterThanThreeMinutes(SimpleDateFormat sdf, String date1, String date2) {
        try {
            Date d1 = sdf.parse(date1);
            Date d2 = sdf.parse(date2);
            long diffInMillies = Math.abs(d2.getTime() - d1.getTime());
            long diffInMinutes = diffInMillies / (1000 * 60);
            return diffInMinutes > 5;
        } catch (Exception e) {
            System.out.println("Erro ao comparar datas: " + e.getMessage());
            return false;
        }
    }
}

class Position {
    String time;
    double latitude;
    double longitude;
    double altitude;
    double groundSpeed;
    String idAircraft;
    String modelAircraft;

    @Override
    public String toString() {
        return "Position{" +
                "time='" + time + '\'' +
                ", latitude=" + latitude +
                ", longitude=" + longitude +
                ", altitude=" + altitude +
                ", groundSpeed=" + groundSpeed +
                ", idAircraft='" + idAircraft + '\'' +
                ", modelAircraft='" + modelAircraft + '\'' +
                '}';
    }
}

class Flight {
    public String saida;
    public String chegada;

    @Override
    public String toString() {
        return "Flight [saida=" + saida + ", chegada=" + chegada + "]";
    }
}
